engine/pkg/dm/message_agent.go

```diff
@@ -0,0 +1,407 @@
+// RequestFuncType(1): func(ctx context.Context, req *interface{}) (resp *interface{}, err error) {}
+// RequestFuncType(2): func(ctx context.Context, req *interface{}) (resp *interface{}) {}
+func NewMessageAgentImpl(id string, commandHandler interface{}, messageHandlerManager p2p.MessageHandlerManager) *MessageAgentImpl {
+	agent := &MessageAgentImpl{
+		messagePair:           newMessagePair(),
+		commandHandler:        commandHandler,
+		messageHandlerManager: messageHandlerManager,
+		pool:                  workerpool.NewDefaultAsyncPool(10),
+		id:                    id,
+	}
+	agent.messageRouter = lib.NewMessageRouter(agent.id, agent.pool, 100,
+		func(topic p2p.Topic, msg p2p.MessageValue) error {
+			err := agent.onMessage(topic, msg)
+			if err != nil {
+				// Todo: handle error
+				log.L().Error("failed to handle message", log.ShortError(err))
+			}
+			return err
+		},
+	)
+	return agent
+}
+
+// Init inits message agent.
+func (agent *MessageAgentImpl) Init(ctx context.Context) error {
+	agent.ctx, agent.cancel = context.WithCancel(context.Background())
+	agent.wg.Add(1)
+	go func() {
+		defer agent.wg.Done()
+		err := agent.pool.Run(agent.ctx)
+		log.L().Info("workerpool exited", zap.Error(err))
+	}()
+	return nil
+}
+
+// Tick implements MessageAgent.Tick
+func (agent *MessageAgentImpl) Tick(ctx context.Context) error {
+	return agent.messageRouter.Tick(ctx)
+}
+
+// Close closes message agent.
+func (agent *MessageAgentImpl) Close(ctx context.Context) error {
+	if agent.cancel != nil {
+		agent.cancel()
+	}
+	agent.wg.Wait()
+	return nil
+}
+
+// UpdateSender adds or deletes the sender by sender-id and register/unregister topic.
+func (agent *MessageAgentImpl) UpdateSender(senderID string, sender Sender) error {
+	if sender == nil {
+		if _, loaded := agent.senders.LoadAndDelete(senderID); loaded {
+			return agent.unregisterTopic(agent.ctx, senderID)
+		}
+	} else if _, loaded := agent.senders.LoadOrStore(senderID, sender); !loaded {
+		return agent.registerTopic(agent.ctx, senderID)
+	}
+	return nil
+}
+
+// getSender gets sender by senderID.
+func (agent *MessageAgentImpl) getSender(senderID string) (Sender, error) {
+	sender, ok := agent.senders.Load(senderID)
+	if !ok {
+		return nil, errors.Errorf("sender %s not found", senderID)
+	}
+	return sender.(Sender), nil
+}
+
+// SendMessage send message asynchronously.
+func (agent *MessageAgentImpl) SendMessage(ctx context.Context, senderID string, command string, msg interface{}) error {
+	sender, err := agent.getSender(senderID)
+	if err != nil {
+		return err
+	}
+	ctx2, cancel := context.WithTimeout(ctx, defaultMessageTimeOut)
+	defer cancel()
+	log.L().Debug("send message", zap.String("sender-id", senderID), zap.String("command", command), zap.Any("msg", msg))
+	return sender.SendMessage(ctx2, generateTopic(agent.id, senderID), message{ID: 0, Type: messageTp, Command: command, Payload: msg}, false /* nonblock */)
```

Comment:
the argument and parameter of generateTopic

agent.id(workerID) -> senderID
senderID -> receiverID

very confusing

Reply:
The reason is that there are two different concepts of senderID, one for the id of [Sender interface](https://github.com/pingcap/tiflow/blob/2ac5f3d75726ebe579fb66e53dd8adfe82e3043c/engine/pkg/dm/message_agent.go#L81) and one for the id of Sender (regsiter/unregister topic), I rename the Sender interface to client now.